<a href="https://colab.research.google.com/github/AndreyPro39/AndreyPro39/blob/main/%D0%9A%D0%BE%D0%BF%D0%B8%D1%8F_%D0%B1%D0%BB%D0%BE%D0%BA%D0%BD%D0%BE%D1%82%D0%B0_%22test_scoring_(4)_(2)_(3)_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [91]:
import pandas as pd
import numpy as np
import sklearn
import imblearn
import matplotlib as plt
from matplotlib import pyplot as plt
!pip install catboost
!pip install xgboost
!pip install lightgbm
!pip install category_encoders
!pip install phik
import phik
from sklearn.model_selection import train_test_split
import seaborn as sns
import datetime
from datetime import datetime
from phik.report import plot_correlation_matrix
from sklearn.preprocessing import LabelEncoder
from category_encoders.cat_boost import CatBoostEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.metrics import roc_auc_score
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.compose import ColumnTransformer
import xgboost as xgb
import lightgbm as lgb
import catboost as cb
!pip install shap -q
import shap
from catboost import CatBoostClassifier
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from scipy.stats import zscore
from phik.report import plot_correlation_matrix
from phik import phik_matrix
from phik.binning import bin_data


In [92]:
from google.colab import drive
drive.mount('/content/drive')
file_path = '/content/drive/My Drive/shift_ml_2025_train.csv'
df_train = pd.read_csv(file_path)


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


/tmp/ipython-input-3860169867.py:4: DtypeWarning: Columns (49) have mixed types. Specify dtype option on import or set low_memory=False.
  df_train = pd.read_csv(file_path)


In [93]:
df_train.head()

id  сумма_займа срок_займа  процентная_ставка  аннуитет рейтинг допрейтинг                           профессия_заемщика     стаж владение_жильем  годовой_доход  подтвержден_ли_доход  итоговый_статус_займа  рейтинг_OCR платежный_график  коэфф_акционных_зачислений_в_остатке                  цель_займа  кол_во_субсидий_за_год  индекс_проживания                регион    пдн пос_стоп_фактор юридический_статус  кол-во_просрочек_за_два_года  коэфф_невыплаченного_сумм_остатка  коэфф_исходного_платежа дата_первого_займа  нижний_порог_рейтинга_заемщика  верхний_порог_рейтинга_заемщика  кол-во_заявок_за_полгода  кол-во_месяцев_с_последней_просрочки  кол-во_месяцев_с_последнего_займа  кол-во_открытых_счетов  кол-во_записей_в_общедоступных_документах  оборотный_баланс  коэфф_загрузки_возобновляемого_счета  кол-во_счетов первоначальный_статус_займа  непогашенная_сумма_из_тела_займов  сумма_выплат_по_просрочкам пени_за_дефолт  дата_следующей_выплаты  грейд_на_детерминаторе  кол-во_взысканий_за_последний_год  кол-во_месяцев_с_последнего_нарушения  код_политики       тип_займа  совокупный_доход_заемщиков  совокупный_пдн_заемщиков совокупный_статус_подтверждения_доходов_заемщиков  кол-во_текущих_просроченных_счетов  общая_сумма_взысканий  возвраты_с_учетом_мпл  общая_сумма_на_счетах  кол-во_открытых_счетов_за_полгода  кол-во_текущих_аннуитетных_счетов  кол-во_аннуитетных_счетов_за_год  кол-во_аннуитетных_счетов_за_2_года  кол-во_месяцев_с_посл_аннуитетного_счета  текущий_баланс_по_аннуитетным_счетам  соотношение_сумм_текущего_баланса_к_лимиту_по_аннуитетным_счетам  кол-во_возоб_счетов_за_год  кол-во_возоб_счетов_за_2_года  макс_баланс_по_возоб_счетам  соотношение_баланса_к_лимиту_общее  общий_лимит_по_возоб_счету  кол-во_финансовых_запросов  кол-во_финансовых_операций  кол-во_заявок_на_кредит_за_год  кол-во_открытых_счетов_за_2_года  средний_баланс_текущих_счетов  суммарная_доступная_сумма_займа_по_картам  соотношение_баланса_к_лимиту_по_картам  кол-во_списаний_за_год  сумма_задолженности  кол-во_месяцев_с_первого_аннуитетного_счета  кол-во_месяцев_с_первого_возобновляемого_счета  кол-во_месяцев_с_последнего_возобновляемого_счета  кол-во_месяцев_с_последнего_счета  кол-во_ипотек  кол-во_месяцев_с_последней_карты  кол-во_мес_с_последней_задолженности_по_карте  кол-во_мес_с_последней_заявки  кол-во_мес_с_последней_задолженности_по_возобновляемому_счету  кол-во_счетов_с_просрочкой_4_мес  кол-во_активных_карт  кол-во_активных_возобновляемых_счетов  кол-во_карт_без_нарушений  кол-во_карт  кол-во_аннуитетных_счетов  кол-во_открытых_возобновляемых_счетов  кол-во_возобновляемых_счетов  кол-во_возобновляемых_счетов_с_балансом_более_0  кол-во_счетов_без_нарушений  кол-во_счетов_с_акт_просрочкой_4_мес_обн_2_мес  кол-во_счетов_с_акт_просрочкой_1_мес  кол-во_счетов_с_акт_просрочкой_3_мес_за_2_года  кол-во_счетов_за_посл_год  процент_счетов_без_просрочек  процент_счетов_прев_75_лимита  кол-во_публ_банкротств  кол-во_залогов  кредитный_лимит  кредитный_баланс_без_ипотеки  лимит_по_картам  лимит_по_аннуитетным_счетам  кредитный_баланс_по_возоб_счетам особая_ситуация тип_предоставления_кредита
0  68355089    1235000.0     3 года              11.99   41014.0       В         В1                                      инженер  10+ лет         ИПОТЕКА      1300000.0        Не подтвержден                      0            4                n                              0.899168               мелкий_бизнес                       0              577.0  Новгородская область  16.06             Нет                Нет                           1.0                       1.345312e+06                 0.767016            12-1999                           715.0                            719.0                       4.0                                   6.0                                NaN                    22.0                                        0.0           21470.0                                  19.2           38.0                           а                             

In [94]:
file_path_2 = '/content/drive/My Drive/shift_ml_2025_test.csv'

In [95]:
df_test = pd.read_csv(file_path_2)

In [96]:
df_test.head()

id  сумма_займа срок_займа  процентная_ставка  аннуитет рейтинг допрейтинг                        профессия_заемщика     стаж владение_жильем  годовой_доход подтвержден_ли_доход  рейтинг_OCR платежный_график  коэфф_акционных_зачислений_в_остатке          цель_займа  кол_во_субсидий_за_год  индекс_проживания                 регион    пдн пос_стоп_фактор юридический_статус  кол-во_просрочек_за_два_года  коэфф_невыплаченного_сумм_остатка  коэфф_исходного_платежа дата_первого_займа  нижний_порог_рейтинга_заемщика  верхний_порог_рейтинга_заемщика  кол-во_заявок_за_полгода  кол-во_месяцев_с_последней_просрочки  кол-во_месяцев_с_последнего_займа  кол-во_открытых_счетов  кол-во_записей_в_общедоступных_документах  оборотный_баланс  коэфф_загрузки_возобновляемого_счета  кол-во_счетов первоначальный_статус_займа  непогашенная_сумма_из_тела_займов  сумма_выплат_по_просрочкам  пени_за_дефолт  дата_следующей_выплаты  грейд_на_детерминаторе  кол-во_взысканий_за_последний_год  кол-во_месяцев_с_последнего_нарушения  код_политики       тип_займа  совокупный_доход_заемщиков  совокупный_пдн_заемщиков совокупный_статус_подтверждения_доходов_заемщиков  кол-во_текущих_просроченных_счетов  общая_сумма_взысканий  возвраты_с_учетом_мпл  общая_сумма_на_счетах  кол-во_открытых_счетов_за_полгода  кол-во_текущих_аннуитетных_счетов  кол-во_аннуитетных_счетов_за_год  кол-во_аннуитетных_счетов_за_2_года  кол-во_месяцев_с_посл_аннуитетного_счета  текущий_баланс_по_аннуитетным_счетам  соотношение_сумм_текущего_баланса_к_лимиту_по_аннуитетным_счетам  кол-во_возоб_счетов_за_год  кол-во_возоб_счетов_за_2_года  макс_баланс_по_возоб_счетам  соотношение_баланса_к_лимиту_общее  общий_лимит_по_возоб_счету  кол-во_финансовых_запросов  кол-во_финансовых_операций  кол-во_заявок_на_кредит_за_год  кол-во_открытых_счетов_за_2_года  средний_баланс_текущих_счетов  суммарная_доступная_сумма_займа_по_картам  соотношение_баланса_к_лимиту_по_картам  кол-во_списаний_за_год  сумма_задолженности  кол-во_месяцев_с_первого_аннуитетного_счета  кол-во_месяцев_с_первого_возобновляемого_счета  кол-во_месяцев_с_последнего_возобновляемого_счета  кол-во_месяцев_с_последнего_счета  кол-во_ипотек  кол-во_месяцев_с_последней_карты  кол-во_мес_с_последней_задолженности_по_карте  кол-во_мес_с_последней_заявки  кол-во_мес_с_последней_задолженности_по_возобновляемому_счету  кол-во_счетов_с_просрочкой_4_мес  кол-во_активных_карт  кол-во_активных_возобновляемых_счетов  кол-во_карт_без_нарушений  кол-во_карт  кол-во_аннуитетных_счетов  кол-во_открытых_возобновляемых_счетов  кол-во_возобновляемых_счетов  кол-во_возобновляемых_счетов_с_балансом_более_0  кол-во_счетов_без_нарушений  кол-во_счетов_с_акт_просрочкой_4_мес_обн_2_мес  кол-во_счетов_с_акт_просрочкой_1_мес  кол-во_счетов_с_акт_просрочкой_3_мес_за_2_года  кол-во_счетов_за_посл_год  процент_счетов_без_просрочек  процент_счетов_прев_75_лимита  кол-во_публ_банкротств  кол-во_залогов  кредитный_лимит  кредитный_баланс_без_ипотеки  лимит_по_картам  лимит_по_аннуитетным_счетам  кредитный_баланс_по_возоб_счетам особая_ситуация тип_предоставления_кредита
0  85540387     450000.0     3 года               9.49   14413.0       Б         Б2                     обслуживание клиентов  10+ лет         ИПОТЕКА       840000.0       Не подтвержден            4                n                              0.458947  консолидация_долга                       0              601.0   Республика Татарстан  19.09             Нет                 Да                           0.0                       1.345312e+06                 0.812613             6-1995                           740.0                            744.0                       0.0                                   NaN                                NaN                    11.0                                        0.0            7735.0                                  34.1           20.0                           а                                0.0                         0.0             NaN                 

Выделим целевой признак.

In [97]:
target = df_train['итоговый_статус_займа']

Данные загружены. Изучим общую информацию о данных.

In [98]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1210779 entries, 0 to 1210778
Columns: 109 entries, id to тип_предоставления_кредита
dtypes: float64(84), int64(6), object(19)
memory usage: 1006.9+ MB


Имеем в тренировочной выборке 1210779 строк и 109 столбцов весом больше 1 Гб. Попробуем уменьшить размерность данных.

In [99]:
def optimize_data(df):
    for col in df.select_dtypes(include=['int64']).columns:
        df[col] = pd.to_numeric(df[col], downcast='integer')
    for col in df.select_dtypes(include=['float64']).columns:
        df[col] = pd.to_numeric(df[col], downcast='float')
    return df

In [100]:
df_train = optimize_data(df_train)

In [101]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1210779 entries, 0 to 1210778
Columns: 109 entries, id to тип_предоставления_кредита
dtypes: float32(74), float64(10), int32(2), int8(4), object(19)
memory usage: 623.5+ MB


Видим, что размерность снизилась с 1Гб до 628 мегабайт. Проделаем то же самое с тестовой выборкой.

In [102]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 134531 entries, 0 to 134530
Columns: 108 entries, id to тип_предоставления_кредита
dtypes: float64(85), int64(5), object(18)
memory usage: 110.9+ MB


In [103]:
df_test = optimize_data(df_test)

In [104]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 134531 entries, 0 to 134530
Columns: 108 entries, id to тип_предоставления_кредита
dtypes: float32(76), float64(9), int32(2), int8(3), object(18)
memory usage: 68.1+ MB


In [105]:
df_train_len = len(df_train)

Обьединим трейн и тест методом merge.

In [106]:
duplicates = df_train.columns[df_train.columns.duplicated()]
print(f"Дублирующиеся столбцы: {duplicates}")

duplicates_test = df_test.columns[df_test.columns.duplicated()]
print(f"Дублирующиеся столбцы: {duplicates_test}")

Дублирующиеся столбцы: Index([], dtype='object')
Дублирующиеся столбцы: Index([], dtype='object')


In [107]:
data = pd.concat([df_train, df_test], ignore_index=True)
data.head()

id  сумма_займа срок_займа  процентная_ставка  аннуитет рейтинг допрейтинг                           профессия_заемщика     стаж владение_жильем  годовой_доход  подтвержден_ли_доход  итоговый_статус_займа  рейтинг_OCR платежный_график  коэфф_акционных_зачислений_в_остатке                  цель_займа  кол_во_субсидий_за_год  индекс_проживания                регион        пдн пос_стоп_фактор юридический_статус  кол-во_просрочек_за_два_года  коэфф_невыплаченного_сумм_остатка  коэфф_исходного_платежа дата_первого_займа  нижний_порог_рейтинга_заемщика  верхний_порог_рейтинга_заемщика  кол-во_заявок_за_полгода  кол-во_месяцев_с_последней_просрочки  кол-во_месяцев_с_последнего_займа  кол-во_открытых_счетов  кол-во_записей_в_общедоступных_документах  оборотный_баланс  коэфф_загрузки_возобновляемого_счета  кол-во_счетов первоначальный_статус_займа  непогашенная_сумма_из_тела_займов  сумма_выплат_по_просрочкам пени_за_дефолт  дата_следующей_выплаты  грейд_на_детерминаторе  кол-во_взысканий_за_последний_год  кол-во_месяцев_с_последнего_нарушения  код_политики       тип_займа  совокупный_доход_заемщиков  совокупный_пдн_заемщиков совокупный_статус_подтверждения_доходов_заемщиков  кол-во_текущих_просроченных_счетов  общая_сумма_взысканий  возвраты_с_учетом_мпл  общая_сумма_на_счетах  кол-во_открытых_счетов_за_полгода  кол-во_текущих_аннуитетных_счетов  кол-во_аннуитетных_счетов_за_год  кол-во_аннуитетных_счетов_за_2_года  кол-во_месяцев_с_посл_аннуитетного_счета  текущий_баланс_по_аннуитетным_счетам  соотношение_сумм_текущего_баланса_к_лимиту_по_аннуитетным_счетам  кол-во_возоб_счетов_за_год  кол-во_возоб_счетов_за_2_года  макс_баланс_по_возоб_счетам  соотношение_баланса_к_лимиту_общее  общий_лимит_по_возоб_счету  кол-во_финансовых_запросов  кол-во_финансовых_операций  кол-во_заявок_на_кредит_за_год  кол-во_открытых_счетов_за_2_года  средний_баланс_текущих_счетов  суммарная_доступная_сумма_займа_по_картам  соотношение_баланса_к_лимиту_по_картам  кол-во_списаний_за_год  сумма_задолженности  кол-во_месяцев_с_первого_аннуитетного_счета  кол-во_месяцев_с_первого_возобновляемого_счета  кол-во_месяцев_с_последнего_возобновляемого_счета  кол-во_месяцев_с_последнего_счета  кол-во_ипотек  кол-во_месяцев_с_последней_карты  кол-во_мес_с_последней_задолженности_по_карте  кол-во_мес_с_последней_заявки  кол-во_мес_с_последней_задолженности_по_возобновляемому_счету  кол-во_счетов_с_просрочкой_4_мес  кол-во_активных_карт  кол-во_активных_возобновляемых_счетов  кол-во_карт_без_нарушений  кол-во_карт  кол-во_аннуитетных_счетов  кол-во_открытых_возобновляемых_счетов  кол-во_возобновляемых_счетов  кол-во_возобновляемых_счетов_с_балансом_более_0  кол-во_счетов_без_нарушений  кол-во_счетов_с_акт_просрочкой_4_мес_обн_2_мес  кол-во_счетов_с_акт_просрочкой_1_мес  кол-во_счетов_с_акт_просрочкой_3_мес_за_2_года  кол-во_счетов_за_посл_год  процент_счетов_без_просрочек  процент_счетов_прев_75_лимита  кол-во_публ_банкротств  кол-во_залогов  кредитный_лимит  кредитный_баланс_без_ипотеки  лимит_по_картам  лимит_по_аннуитетным_счетам  кредитный_баланс_по_возоб_счетам особая_ситуация тип_предоставления_кредита
0  68355089    1235000.0     3 года              11.99   41014.0       В         В1                                      инженер  10+ лет         ИПОТЕКА      1300000.0        Не подтвержден                    0.0            4                n                              0.899168               мелкий_бизнес                       0              577.0  Новгородская область  16.059999             Нет                Нет                           1.0                       1.345312e+06                 0.767016            12-1999                           715.0                            719.0                       4.0                                   6.0                                NaN                    22.0                                        0.0           21470.0                             19.200001           38.0                           а                     

In [108]:
data.shape

(1345310, 109)

Найдем пропуски и явные дубликаты в таблицах.

In [109]:
pd.set_option('display.max_colwidth', 110)

In [110]:
data.duplicated().sum()

np.int64(0)

Дубликатов нет.

Выведем все названия столбцов трейна на экран.

Уберем ограничение максимального количества отображения колонок в одной ячейке.

In [111]:
pd.set_option('display.max_column', None)
pd.set_option('display.expand_frame_repr', False)
sorted_cols = data.columns.sort_values()
print(sorted_cols.to_list())

['id', 'аннуитет', 'верхний_порог_рейтинга_заемщика', 'владение_жильем', 'возвраты_с_учетом_мпл', 'годовой_доход', 'грейд_на_детерминаторе', 'дата_первого_займа', 'дата_следующей_выплаты', 'допрейтинг', 'индекс_проживания', 'итоговый_статус_займа', 'код_политики', 'кол-во_активных_возобновляемых_счетов', 'кол-во_активных_карт', 'кол-во_аннуитетных_счетов', 'кол-во_аннуитетных_счетов_за_2_года', 'кол-во_аннуитетных_счетов_за_год', 'кол-во_взысканий_за_последний_год', 'кол-во_возоб_счетов_за_2_года', 'кол-во_возоб_счетов_за_год', 'кол-во_возобновляемых_счетов', 'кол-во_возобновляемых_счетов_с_балансом_более_0', 'кол-во_залогов', 'кол-во_записей_в_общедоступных_документах', 'кол-во_заявок_за_полгода', 'кол-во_заявок_на_кредит_за_год', 'кол-во_ипотек', 'кол-во_карт', 'кол-во_карт_без_нарушений', 'кол-во_мес_с_последней_задолженности_по_возобновляемому_счету', 'кол-во_мес_с_последней_задолженности_по_карте', 'кол-во_мес_с_последней_заявки', 'кол-во_месяцев_с_первого_аннуитетного_счета', 'ко

На первый взгляд видно, что очень много признаков и лишних тоже много.\
Если взять из практики модель кредитного скоринга, то можно сразу убрать лишние признаки,\
такие как грейд_на_детерминаторе например и количество_возобновляемых_счетов, доп рейтинг и так далее.

Посмотрим на уникальные значения каждой колонки в отдельности.

In [112]:
for col in data.columns:
    columns_unique = data[col].sort_values().unique()
    print(f'уникальные значения столбца {col}: {columns_unique}')

уникальные значения столбца id: [    54734     55716     55742 ... 145623694 145627026 145636374]
уникальные значения столбца сумма_займа: [  25000.   35000.   36250. ... 1997500. 1998750. 2000000.]
уникальные значения столбца срок_займа: ['3 года' '5 лет']
уникальные значения столбца процентная_ставка: [ 5.31  5.32  5.42  5.79  5.93  5.99  6.    6.03  6.07  6.08  6.11  6.17
  6.19  6.24  6.39  6.46  6.49  6.54  6.62  6.67  6.68  6.71  6.72  6.76
  6.83  6.89  6.91  6.92  6.97  6.99  7.02  7.05  7.07  7.12  7.14  7.21
  7.24  7.26  7.29  7.34  7.35  7.37  7.39  7.4   7.42  7.43  7.46  7.49
  7.51  7.56  7.59  7.62  7.66  7.68  7.69  7.74  7.75  7.84  7.88  7.89
  7.9   7.91  7.96  7.97  7.99  8.    8.07  8.08  8.18  8.19  8.24  8.32
  8.38  8.39  8.46  8.49  8.59  8.6   8.63  8.67  8.7   8.81  8.88  8.9
  8.94  8.99  9.01  9.07  9.16  9.17  9.2   9.25  9.32  9.33  9.38  9.43
  9.44  9.45  9.49  9.51  9.58  9.62  9.63  9.64  9.67  9.7   9.71  9.75
  9.76  9.8   9.83  9.88  9.91  9.92  9

**Вывод: Данные требуют обработки. Есть пропуски, есть лог запись данных, требующая нормализации. Некоторые столбцы не содержат данных,\
либо содержат всего одно значение. Удалим такие столбцы.**

На основе анализа списка параметров и согласно интернету, следующие колонки маловероятно влияют на модель кредитного скоринга или имеют минимальное влияние:\
Технические параметры системы: код_политики, грейд_на_детерминаторе, дата_следующей_выплаты

Избыточные или дублирующие показатели:
совокупный_доход_заемщиков (дублируется с годовой_доход),
совокупный_пдн_заемщиков (дублируется с пдн),
кол-во_заявок_за_полгода (частично дублируется с другими метриками активности).

Показатели с низкой предсказательной силой:
кол-во_публичных_банкротств (редко встречается в данных),
кол-во_залогов (не всегда коррелирует с платежеспособностью),
коэфф_акционных_зачислений_в_остатке (не связан напрямую с платежеспособностью).

Промежуточные расчетные показатели:
оборотный_баланс,
коэфф_загрузки_возобновляемого_счета,
возвраты_с_учетом_мпл.

Показатели с высокой корреляцией с другими метриками:
кол-во_открытых_счетов_за_полгода (коррелирует с общей активностью),
кол-во_финансовых_операций (частично дублирует активность),
процент_счетов_прев_75_лимита (уже учтен в других метриках лимитов).
Удалим эти признаки для дальнейшей оптимизации данных для обучения модели.

In [113]:
data = data.drop(
    ['код_политики',  'непогашенная_сумма_из_тела_займов', 'грейд_на_детерминаторе', 'дата_следующей_выплаты', 'совокупный_доход_заемщиков',
'совокупный_пдн_заемщиков', 'кол-во_заявок_за_полгода', 'кол-во_публ_банкротств', 'кол-во_залогов', 'платежный_график',
'коэфф_акционных_зачислений_в_остатке', 'оборотный_баланс', 'коэфф_загрузки_возобновляемого_счета',
'возвраты_с_учетом_мпл', 'кол-во_открытых_счетов_за_полгода', 'кол-во_финансовых_операций', 'процент_счетов_прев_75_лимита'], axis=1
)

In [114]:
data.shape

(1345310, 92)

Минусовали по 17 лишних столбцов обьединенного датафрейма.

Посмотрим на пропуски после удаления лишних признаков.

In [115]:
pd.set_option('display.max_rows', None)
data.isna().sum().sort_values(ascending=False)

кредитный_баланс_по_возоб_счетам                                    1326681
совокупный_статус_подтверждения_доходов_заемщиков                   1319715
кол-во_месяцев_с_последнего_займа                                   1116755
кол-во_мес_с_последней_задолженности_по_карте                       1026290
кол-во_месяцев_с_последнего_нарушения                                991560
кол-во_мес_с_последней_задолженности_по_возобновляемому_счету        895348
соотношение_сумм_текущего_баланса_к_лимиту_по_аннуитетным_счетам     880294
кол-во_месяцев_с_посл_аннуитетного_счета                             821928
соотношение_баланса_к_лимиту_общее                                   807765
кол-во_заявок_на_кредит_за_год                                       807713
текущий_баланс_по_аннуитетным_счетам                                 807712
кол-во_аннуитетных_счетов_за_2_года                                  807712
кол-во_текущих_аннуитетных_счетов                                    807712
кол-во_возоб_счетов_за_год                                           807712
кол-во_возоб_счетов_за_2_года                                        807712
кол-во_аннуитетных_счетов_за_год                                     807712
макс_баланс_по_возоб_счетам                                          807712
кол-во_финансовых_запросов                                           807712
кол-во_месяцев_с_последней_просрочки                                 678743
пени_за_дефолт                                                       497972
кол-во_мес_с_последней_заявки                                        174071
итоговый_статус_займа                                                134531
кол-во_счетов_с_акт_просрочкой_4_мес_обн_2_мес                       117401
кол-во_месяцев_с_первого_аннуитетного_счета                          105575
профессия_заемщика                                                    85785
стаж                                                                  78511
процент_счетов_без_просрочек                                          67681
средний_баланс_текущих_счетов                                         67549
кол-во_месяцев_с_последнего_возобновляемого_счета                     67528
кол-во_возобновляемых_счетов                                          67528
кол-во_месяцев_с_первого_возобновляемого_счета                        67528
кол-во_открытых_возобновляемых_счетов                                 67527
кол-во_счетов_с_акт_просрочкой_3_мес_за_2_года                        67527
кол-во_счетов_с_акт_просрочкой_1_мес                                  67527
кол-во_возобновляемых_счетов_с_балансом_более_0                       67527
кол-во_счетов_с_просрочкой_4_мес                                      67527
лимит_по_аннуитетным_счетам                                           67527
кол-во_счетов_за_посл_год                                             67527
кредитный_лимит                                                       67527
общая_сумма_взысканий                                                 67527
кол-во_активных_карт                                                  67527
кол-во_аннуитетных_счетов                                             67527
общая_сумма_на_счетах                                                 67527
общий_лимит_по_возоб_счету                                            67527
кол-во_активных_возобновляемых_счетов                                 67527
кол-во_карт                                                           67527
кол-во_месяцев_с_последнего_счета                                     67527
соотношение_баланса_к_лимиту_по_картам                                61912
суммарная_доступная_сумма_займа_по_картам                             61143
кол-во_месяцев_с_последней_карты                                      60221
кол-во_карт_без_нарушений                                             55841
кол-во_счетов_без_нарушений                                           55841
кредитный_баланс_без_ипотеки                    

Обработаем пропуски. Такое большое количество пропусков в данных столбцах не означает что данные нужно удалять.\
Если удалим столько записей, будет большая потеря данных. Скорее нужно заменить данные пропуски значениями-заглушками, потомучто
модель должна работать со всеми типами клиентов. Тем более в нескольких колонках количество пропусков одинаково - значит и не было там никаких значений.

Например в столбце кредитный_баланс_по_возоб_счетам - более 98% пропусков,\
что связано скорее всего с тем, что клиенты не пользовались возобновляемыми счетами повторно.\
Поэтому заполним данные пропуски нулями. Аналогично сделаем с подобными пропусками.

In [116]:
columns = [
'кредитный_баланс_по_возоб_счетам', 'кол-во_месяцев_с_последнего_займа', 'кол-во_мес_с_последней_задолженности_по_карте',
'кол-во_месяцев_с_последнего_нарушения', 'кол-во_мес_с_последней_задолженности_по_возобновляемому_счету',
'соотношение_сумм_текущего_баланса_к_лимиту_по_аннуитетным_счетам', 'кол-во_месяцев_с_посл_аннуитетного_счета',
'соотношение_баланса_к_лимиту_общее', 'кол-во_заявок_на_кредит_за_год', 'кол-во_возоб_счетов_за_год',
'кол-во_возоб_счетов_за_2_года', 'кол-во_аннуитетных_счетов_за_год', 'кол-во_текущих_аннуитетных_счетов',
'макс_баланс_по_возоб_счетам', 'кол-во_финансовых_запросов', 'текущий_баланс_по_аннуитетным_счетам',
'кол-во_месяцев_с_последней_просрочки', 'процент_счетов_без_просрочек', 'средний_баланс_текущих_счетов',
'кол-во_месяцев_с_первого_возобновляемого_счета', 'кол-во_месяцев_с_последнего_возобновляемого_счета',
'кол-во_возобновляемых_счетов', 'кредитный_лимит', 'общий_лимит_по_возоб_счету',
'кол-во_счетов_за_посл_год',
'кол-во_счетов_с_акт_просрочкой_3_мес_за_2_года',
'кол-во_счетов_с_акт_просрочкой_1_мес',
'общая_сумма_на_счетах',
'кол-во_карт',
'лимит_по_аннуитетным_счетам',
'кол-во_аннуитетных_счетов',
'общая_сумма_взысканий',
'кол-во_открытых_возобновляемых_счетов',
'кол-во_счетов_с_просрочкой_4_мес',
'кол-во_месяцев_с_последнего_счета',
'кол-во_возобновляемых_счетов_с_балансом_более_0',
'кол-во_активных_возобновляемых_счетов',
'кол-во_активных_карт',
'соотношение_баланса_к_лимиту_по_картам',
'суммарная_доступная_сумма_займа_по_картам',
'кол-во_месяцев_с_последней_карты',
'кол-во_карт_без_нарушений',
'кол-во_счетов_без_нарушений',
'лимит_по_картам',
'кредитный_баланс_без_ипотеки',
'кол-во_открытых_счетов_за_2_года',
'кол-во_ипотек',
'пдн',
'кол-во_взысканий_за_последний_год',
'кол-во_списаний_за_год',
'пени_за_дефолт'
]
data[columns] = data[columns].fillna(0)

Нормализуем столбцы с лог записью.

In [117]:
numeric_cols = data.select_dtypes(include=['float64', 'float32', 'int64', 'int32']).columns
data[numeric_cols] = (data[numeric_cols] - data[numeric_cols].min()) / \
                  (data[numeric_cols].max() - data[numeric_cols].min())

In [118]:
data.head()

id  сумма_займа срок_займа  процентная_ставка  аннуитет рейтинг допрейтинг                           профессия_заемщика     стаж владение_жильем  годовой_доход  подтвержден_ли_доход  итоговый_статус_займа  рейтинг_OCR                  цель_займа  кол_во_субсидий_за_год  индекс_проживания                регион      пдн пос_стоп_фактор юридический_статус  кол-во_просрочек_за_два_года  коэфф_невыплаченного_сумм_остатка  коэфф_исходного_платежа дата_первого_займа  нижний_порог_рейтинга_заемщика  верхний_порог_рейтинга_заемщика  кол-во_месяцев_с_последней_просрочки  кол-во_месяцев_с_последнего_займа  кол-во_открытых_счетов  кол-во_записей_в_общедоступных_документах  кол-во_счетов первоначальный_статус_займа  сумма_выплат_по_просрочкам пени_за_дефолт  кол-во_взысканий_за_последний_год  кол-во_месяцев_с_последнего_нарушения       тип_займа совокупный_статус_подтверждения_доходов_заемщиков  кол-во_текущих_просроченных_счетов  общая_сумма_взысканий  общая_сумма_на_счетах  кол-во_текущих_аннуитетных_счетов  кол-во_аннуитетных_счетов_за_год  кол-во_аннуитетных_счетов_за_2_года  кол-во_месяцев_с_посл_аннуитетного_счета  текущий_баланс_по_аннуитетным_счетам  соотношение_сумм_текущего_баланса_к_лимиту_по_аннуитетным_счетам  кол-во_возоб_счетов_за_год  кол-во_возоб_счетов_за_2_года  макс_баланс_по_возоб_счетам  соотношение_баланса_к_лимиту_общее  общий_лимит_по_возоб_счету  кол-во_финансовых_запросов  кол-во_заявок_на_кредит_за_год  кол-во_открытых_счетов_за_2_года  средний_баланс_текущих_счетов  суммарная_доступная_сумма_займа_по_картам  соотношение_баланса_к_лимиту_по_картам  кол-во_списаний_за_год  сумма_задолженности  кол-во_месяцев_с_первого_аннуитетного_счета  кол-во_месяцев_с_первого_возобновляемого_счета  кол-во_месяцев_с_последнего_возобновляемого_счета  кол-во_месяцев_с_последнего_счета  кол-во_ипотек  кол-во_месяцев_с_последней_карты  кол-во_мес_с_последней_задолженности_по_карте  кол-во_мес_с_последней_заявки  кол-во_мес_с_последней_задолженности_по_возобновляемому_счету  кол-во_счетов_с_просрочкой_4_мес  кол-во_активных_карт  кол-во_активных_возобновляемых_счетов  кол-во_карт_без_нарушений  кол-во_карт  кол-во_аннуитетных_счетов  кол-во_открытых_возобновляемых_счетов  кол-во_возобновляемых_счетов  кол-во_возобновляемых_счетов_с_балансом_более_0  кол-во_счетов_без_нарушений  кол-во_счетов_с_акт_просрочкой_4_мес_обн_2_мес  кол-во_счетов_с_акт_просрочкой_1_мес  кол-во_счетов_с_акт_просрочкой_3_мес_за_2_года  кол-во_счетов_за_посл_год  процент_счетов_без_просрочек  кредитный_лимит  кредитный_баланс_без_ипотеки  лимит_по_картам  лимит_по_аннуитетным_счетам  кредитный_баланс_по_возоб_счетам особая_ситуация тип_предоставления_кредита
0  0.469155     0.612658     3 года           0.260125  0.475450       В         В1                                      инженер  10+ лет         ИПОТЕКА       0.005910        Не подтвержден                    0.0            4               мелкий_бизнес                       0           0.574597  Новгородская область  0.01706             Нет                Нет                      0.025641                                NaN                 0.767018            12-1999                        0.409091                         0.407240                              0.026549                                0.0                0.244444                                        0.0       0.206897                           а                4.289932e-12          False                                0.0                               0.000000  Индивидуальный                                               NaN                                 0.0                    0.0               0.025549                           0.017544                               0.0                             0.019608                                  0.037182                              0.010523                                                          0.130824                    0.071429                       0.056604                     0.008331  

In [ ]:
data['профессия_заемщика']

In [ ]:
data.shape

In [ ]:
data.isna().sum()

все еще остались пропуски. Обработаем их.

In [ ]:
data['профессия_заемщика']

In [ ]:
if 'профессия_заемщика' in data.columns:
    data['профессия_заемщика'] = data['профессия_заемщика'].fillna('нет профессии')
else:
    print("Столбец 'профессия_заемщика' отсутствует в датафрейме")


In [ ]:
data['стаж'].value_counts()

расчитаем средний стаж для заполнения пропусков.

In [ ]:
# Создаем словарь с данными
data = {
    'стаж': {
        '10+ лет': 442199,
        '2 года': 121743,
        '< 1 года': 108061,
        '3 года': 107597,
        '1 год': 88494,
        '5 лет': 84154,
        '4 года': 80556,
        '6 лет': 62733,
        '8 лет': 60701,
        '7 лет': 59624,
        '9 лет': 50937
    }
}

# Преобразуем категории стажа в числовые значения
def convert_experience(exp):
    if exp == '< 1 года':
        return 0.5
    elif exp == '10+ лет':
        return 10
    else:
        return int(exp.split()[0])

# Рассчитываем средний стаж
total_experience = 0
total_count = 0

for exp, count in data['стаж'].items():
    numeric_exp = convert_experience(exp)
    total_experience += numeric_exp * count
    total_count += count

average_experience = total_experience / total_count

print(f"Средний стаж: {average_experience:.2f} лет")


In [ ]:
if 'профессия_заемщика' in data.columns:
    data['профессия_заемщика'] = data['профессия_заемщика'].fillna('нет профессии')
else:
    print("Столбец 'профессия_заемщика' отсутствует в датафрейме")


In [ ]:
data['профессия_заемщика'] = data['профессия_заемщика'].fillna('нет профессии')
data['стаж'] = data['стаж'].fillna(average_experience)
data['совокупный_статус_подтверждения_доходов_заемщиков',
'кол-во_аннуитетных_счетов_за_2_года',
'кол-во_месяцев_с_первого_аннуитетного_счета',
'кол-во_мес_с_последней_заявки',
'кол-во_счетов_с_акт_просрочкой_4_мес_обн_2_мес'
].fillna(0)








100% пропусков в столбце 'коэфф_невыплаченного_сумм_остатка'. Удалим данный столбец.

In [ ]:
data[numeric_cols].drop(['id', 'коэфф_невыплаченного_сумм_остатка'], axis=1)

In [ ]:
def columns_unique(data, numeric_cols):
available_cols = [col for col in numeric_cols if col in data.columns]

    for col in available_cols:
        plt.figure(figsize=(10, 5))

        # Проверяем, что столбец содержит числовые данные
        if pd.api.types.is_numeric_dtype(data[col]):
            data[col].hist(bins=30, grid=True, edgecolor='black')

            # Добавляем подписи и заголовок
            plt.xlabel(col, fontsize=12)
            plt.ylabel('Количество', fontsize=12)
            plt.title(f'Распределение признака: {col}', fontsize=14)

            # Добавляем сетку и улучшаем оформление
            plt.grid(axis='y', alpha=0.75)
            plt.tight_layout()
            plt.show()
        else:
            print(f"Предупреждение: столбец {col} не является числовым")

In [ ]:
columns_unique(data, numeric_cols)

In [ ]:
cat_cols = data.select_dtypes(include=['object']).columns
cat_cols

In [ ]:
for col in cat_cols:
    plt.figure(figsize=(10, 6))
    sns.boxplot(x='col', y='количество', data=data)
    plt.xlabel(col)
    plt.ylabel('количество')
    plt.grid(True)
    plt.title(f'бокс плот для колонки {col}')
    plt.show()

In [ ]:
# Разделение обратно на тренировочный и тестовый наборы
train_processed = data[:train_len]
test_processed = _data[train_len:]

# Выделение признаков и целевого признака
X_train = train_processed.drop(['target', 'id'], axis=1)
y_train = train_processed['target']
X_test = test_processed.drop(['id'], axis=1)


In [ ]:
data.query('')

In [ ]:
data[numeric_cols].drop(['id', 'коэфф_невыплаченного_сумм_остатка'], axis=1).corr()

In [ ]:
Phik_matrix = data.phik_matrix()
plt.figure(figsize=(14, 12))
sns.heatmap(
    Phik_matrix,
    vmin=0, vmax=1,
    cmap='coolwarm',
    annot=False,
    fmt=".2f",
    linewidths=0.1
)

plt.title('Матрица корреляций phik_matrix'
plt.xticks(rotation=90)
plt.yticks(rotation=0)
plt.tight_layout()
plt.show()

In [ ]:
sorted_cols = df_test.columns.sort_values()
print(sorted_cols.to_list())

Пропуски обработали, явных и неявных дубликатов не нашли. Построим графики признаков для наглядности\
и выявления выбросов и аномалий. Для этого создадим функцию с циклом для построения гистограмм.

In [ ]:
def columns_unique(data):
    for col in data.columns:
        plt.figure(figsize=(10, 5))
        data[col].hist(grid=True)
        plt.xlabel(col)
        plt.ylabel('Количество')
        plt.title(f'Распределение признака: {col}')
        plt.show()

In [ ]:
columns_unique(df_train)

In [ ]:
df_train.corr(method=spearman)

In [ ]:
columns_unique(df_test)

Для оптимизации признаков постороим матрицу корреляций по числовым значениям df.corr()